### ISSUE TITLE
请问微调clue-tok需要多大显存， V100 32GB显存都不够，感谢

### ISSUE THREAD

**Issue by ryanran92** [ghosthamlet/gpt2-ml-torch#6]

**Comment by ghosthamlet** [comment]
用一个GPU确实不够，需要用https://github.com/microsoft/DeepSpeed 这个类库做CPU offload，教程在：https://www.deepspeed.ai/tutorials/zero-offload/ ，我已经试验过，11GB显存的2080ti都可以微调成功。

### 更新：
不过你的GPU有32GB内存，可以先尽可能减小batch大小和seq长度，还是不行再用混合精度训练：https://pytorch.org/docs/stable/notes/amp_examples.html ，仍然不行再用DeepSpeed。
另外即使微调可以顺利进行，收敛也不太容易，lr和weight_decay配置影响很大，我的微调配置是，数据20万以上，epochs：1，batch_size：1，seq_len：300，lr：5e-8，weight_decay：0.01，warmup_steps: 200

**Comment by ryanran92** [comment]
> 用一个GPU确实不够，需要用https://github.com/microsoft/DeepSpeed 这个类库做CPU offload，教程在：https://www.deepspeed.ai/tutorials/zero-offload/ ，我已经试验过，11GB显存的2080ti都可以微调成功。
> 
> ### 更新：
> 不过你的GPU有32GB内存，可以先尽可能减小batch大小和seq长度，还是不行再用混合精度训练：https://pytorch.org/docs/stable/notes/amp_examples.html ，仍然不行再用DeepSpeed。
> 另外即使微调可以顺利进行，收敛也不太容易，lr和weight_decay配置影响很大，我的微调配置是，数据20万以上，epochs：1，batch_size：1，seq_len：300，lr：5e-8，weight_decay：0.01，warmup_steps: 200

ok，非常感谢你的回复。
  1） 我上周尝试了减小batch到1，依然超显存；也尝试了apex混合精度训练，也有问题；其次现在问题可能出现在seq长度，在查看每个batch的数据时，发现每次超显存的时机是在每个batch计算attention时，attn_outputs维度为 （bs，_，1024,1024），这里的config.n_ctx=1024，这个值是模型中固定参数好像无法修改？ 亦或是我没有修改对seqlen（我修改的是full_tokenizer的max_model_len）？不知你seqlen修改的是何处？ 2） 你提到的CPU offload我会在解决seqlen问题之后会进行尝试，是个很棒的想法（至少我以前没有尝试过）。
   再次感谢，并希望得到1问题的后续回复。

**Comment by ghosthamlet** [comment]
1) seq_len是你构建Datasets时控制的，修改为比预训练模型更小的话，不需要修改模型和tokenizer配置。如果seq_len已经300以下但还是不行，可能是你的混合精度做的有问题，新版pytorch已经内置混合精度，可能比apex做更容易点，如果确定混合精度没有问题，那只能用DeepSpeed。

**Comment by ryanran92** [comment]
> 1. seq_len是你构建Datasets时控制的，修改为比预训练模型更小的话，不需要修改模型和tokenizer配置。如果seq_len已经300以下但还是不行，可能是你的混合精度做的有问题，新版pytorch已经内置混合精度，可能比apex做更容易点，如果确定混合精度没有问题，那只能用DeepSpeed。

ok, 了解了，多谢

**Comment by ghosthamlet** [comment]
不客气

**Comment by hsiehjackson** [comment]
Hi, can you share your DeepSpeed config setting? I want to fine-tune; however, it seems that calling model.half() by using fp16 in deepspeed will cause performance drop. Thanks!

**Comment by ghosthamlet** [comment]
@hsiehjackson 
My deepspeed config:
 {
        'zero_optimization': {
            'stage': 2,
            'cpu_offload': True,
            'contiguous_gradients': True,
            # see: https://github.com/microsoft/DeepSpeed/issues/467
            'overlap_comm': False,
            'reduce_bucket_size': 3000000,
            'allgather_bucket_size': 3000000
            },
        'train_batch_size': args.batch_size * args.world_size, // 1 * 4
        "gradient_accumulation_steps": args.gradient_accumulation_steps, // 1
        'fp16': {
            'enabled': True,
            "loss_scale": 0,
            "loss_scale_window": 1000,
            "hysteresis": 2,
            "min_loss_scale": 1,
            },
          "activation_checkpointing": {
            "partition_activations": True,
            "contiguous_memory_optimization": True,
            "cpu_checkpointing": True
          }, 
         "wall_clock_breakdown": False,
}   

I find the more important config that causes performance drop when I finetuning is lr, lr should be very small: 5e-8, and if weight_decay is large like 0.9 then the convergence is very difficult，so my weight_decay is 0.01，and the warmup_steps is 200.

By the way, if the finetune performance drop is not large, e.g. loss begins with 3, ending around 3.3, then it should be ok.

**Comment by hsiehjackson** [comment]
Thanks for your sharing!

**Comment by ghosthamlet** [comment]
You are welcome.

**Comment by ghosthamlet** [comment]
@ranyu @hsiehjackson 
### uploaded finetune text generation code：https://github.com/ghosthamlet/gpt2-ml-torch/blob/master/finetune_lm.py , finetune problem please create new issue.
